### PR TITLE
[TIMOB-26312]: iOS 12 Expose new NSUserActivity APIs for Siri Intents

### PR DESCRIPTION
**Pull request by vijaysingh-axway**
https://jira.appcelerator.org/browse/TIMOB-26312

### CODE REVIEW

File: iphone/Classes/TiAppiOSUserActivityProxy.m [hunk 1]
```diff
@@ -425,6 +438,73 @@ - (void)resignCurrent:(id)unused
...
+  ENSURE_TYPE(value, NSString);
+  if ([TiUtils isIOSVersionLower:@"12.0"]) {
+    return;
+  }
+  [_userActivity setPersistentIdentifier:[TiUtils stringValue:value]];
+}
+
+- (void)deleteSavedUserActivitiesForPersistentIdentifiers:(id)persistentIdentifiers
+{
+  ENSURE_ARRAY(persistentIdentifiers);
```
Comment (hansemannn) [inline iphone/Classes/TiAppiOSUserActivityProxy.m]: Did you actually test all the new API's so far? Since this is a method, you need to unbox the `persistentIdentifiers` attribute with `ENSURE_SINGLE_ARG(persistentIdentifiers, NSArray)`.

File: iphone/Classes/TiAppiOSUserActivityProxy.m [hunk 2]
```diff
@@ -425,6 +438,73 @@ - (void)resignCurrent:(id)unused
...
+    return NUMBOOL(NO);
+  }
+
+  return NUMBOOL(_userActivity.isEligibleForPrediction);
+}
+
+- (void)setEligibleForPrediction:(NSNumber *)value
+{
+  ENSURE_TYPE(value, NSNumber);
+  ENSURE_UI_THREAD(setEligibleForSearch, value);
```
Comment (hansemannn) [inline iphone/Classes/TiAppiOSUserActivityProxy.m]: The UI-thread should be validated before the type-checks.

File: iphone/Classes/TiAppiOSUserActivityProxy.m [hunk 3]
```diff
@@ -425,6 +438,73 @@ - (void)resignCurrent:(id)unused
...
+
+#if IS_XCODE_10
+
+- (NSNumber *)eligibleForPrediction
+{
+  if (![TiUtils isIOSVersionLower:@"12.0"]) {
+    return NUMBOOL(NO);
+  }
+
+  return NUMBOOL(_userActivity.isEligibleForPrediction);
```
Comment (hansemannn) [inline iphone/Classes/TiAppiOSUserActivityProxy.m]: We should move all these to native Obj-C 2.0 statements, e.g. `@(_userActivity.isEligibleForPrediction)` and `@(NO)`. But that should be done with an own PR and an own ticket. Maybe we can deprecate the old macros in SDK 8.0.0?.

File: apidoc/Titanium/App/iOS/UserActivity.yml [hunk 4]
```diff
@@ -69,6 +69,29 @@ properties:
     default: true
     availability: creation
 
+  - name: eligibleForPrediction
+    summary: |
+        A Boolean value that determines whether Siri can suggest the user activity as a shortcut to the user.
+    description: |
+        To donate a user activity to Siri Shortcuts, set eligibleForPrediction to 'true' and make the 
+        user activity current. To make the user activity current, call the becomeCurrent method of activity.
+        For more information, see https://developer.apple.com/documentation/sirikit/donating_shortcuts?language=objc.
```
Comment (hansemannn) [inline apidoc/Titanium/App/iOS/UserActivity.yml]:
```yml
For more information, see [the Apple docs](https://developer.apple.com/documentation/sirikit/donating_shortcuts?language=objc).
```

File: apidoc/Titanium/App/iOS/UserActivity.yml [hunk 5]
```diff
@@ -69,6 +69,29 @@ properties:
     default: true
     availability: creation
 
+  - name: eligibleForPrediction
+    summary: |
+        A Boolean value that determines whether Siri can suggest the user activity as a shortcut to the user.
+    description: |
+        To donate a user activity to Siri Shortcuts, set eligibleForPrediction to 'true' and make the
```
Comment (hansemannn) [inline apidoc/Titanium/App/iOS/UserActivity.yml]:
```
`true`
```

File: iphone/Classes/TiAppiOSUserActivityProxy.m [hunk 6]
```diff
@@ -425,6 +438,73 @@ - (void)resignCurrent:(id)unused
   }
   [_userActivity resignCurrent];
 }
+
+#if IS_XCODE_10
+
+- (NSNumber *)eligibleForPrediction
```
Comment (hansemannn) [inline iphone/Classes/TiAppiOSUserActivityProxy.m]: Can you make sure all public API's are also in the interface header? It makes it easier to maintain.

File: iphone/Classes/TiAppiOSUserActivityProxy.h [hunk 7]
```diff
@@ -16,6 +16,13 @@
 - (id)initWithOptions:(NSDictionary *)props;
 @property (nonatomic, strong) NSUserActivity *userActivity;
 
+#if IS_XCODE_10
+- (void)setEligibleForPrediction:(NSNumber *)value;
+- (void)setPersistentIdentifier:(NSString *)value;
+- (void)deleteSavedUserActivitiesForPersistentIdentifiers:(id)persistentIdentifiers;
+- (void)deleteAllSavedUserActivities:(id)unused;
```
Comment (hansemannn) [inline iphone/Classes/TiAppiOSUserActivityProxy.h]: Not only the new methods but all methods ;-). Including code-docs and params-descriptions.

Comment (vijaysingh-axway) [reply iphone/Classes/TiAppiOSUserActivityProxy.h]: All method exposing is fine. Do we use code-docs and params-description in header file in Titanium SDK?

**Comment by sgtcoolguy** [comment]
7_4_X version is #10292

File: iphone/Classes/TiAppiOSUserActivityProxy.h (hunk 7 above)
Comment (hansemannn) [reply iphone/Classes/TiAppiOSUserActivityProxy.h]: For Ti core-developers, it will help knowing which params are used in the "boxed" arguments. But it looks fine for now.

**Review by hansemannn — APPROVED** [review]
